### ISSUE TITLE
Anchors in env key for .travis.yml configs display the wrong values

### ISSUE THREAD

**Issue by Snaipe** [travis-ci/travis-ci#5839]
~~Currently, YAML anchors cannot be used for `env` lists as the final list does not get flattened.~~

~~Flattening it would allow us to make configurations like [this](https://github.com/Snaipe/Mimick/blob/e03ab50f348c5df7e4d9866dd988112f0e46ee05/.travis.yml#L48), which would greatly ease the pain of maintaining complex build matrices.~~

Edit: Actually, nevermind, it does -- it just displays that there are invalid secure variables.

Incidentally, the configuration as it is gets accepted, but expanded values are considered to be secure variables of an invalid format (see [relevant build](https://travis-ci.org/Snaipe/Mimick/builds/119662760)).

**Comment by BanzaiMan** [comment]
Which jobs show there are invalid secure variables? There are no secure variables defined in https://raw.githubusercontent.com/Snaipe/Mimick/e03ab50f348c5df7e4d9866dd988112f0e46ee05/.travis.yml, or the Settings panel.

**Comment by Snaipe** [comment]
@BanzaiMan There are none, but the linked build reports that there are.

**Comment by BanzaiMan** [comment]
I need help locating that information. On which lines in https://travis-ci.org/Snaipe/Mimick/jobs/119662773 and https://travis-ci.org/Snaipe/Mimick/jobs/119662769 do you see that? Searching for "secure" yields nothing.

**Comment by Snaipe** [comment]
Ah, right. I wasn't precise in my report.

What I see is that, on the build summary screen, that environment variables expanded by anchor values get interpreted as secure variables when they aren't.

A screenshot might tell more than words:

![travis-secure-bug](https://cloud.githubusercontent.com/assets/977973/14271111/977decbc-faf5-11e5-89d3-5f4aab3fb631.png)

Here, on the summary line for job 124.4, I would expect the environment line to be "CONFIGURATION=Debug, TOOLCHAIN=gcc-i686" (ditto for jobs 124.5-124.9 & 124.13-124.15)

Interestingly, jobs 124.7-9 succeed in parsing `CMAKE_OPTS` as a secure variable (which it is not), and ignore the rest (`ARCH`, `TARGET` & `TOOLCHAIN`).

**Comment by BanzaiMan** [comment]
I see that now. Thanks. The problem lies with the `.travis.yml`.

In https://github.com/Snaipe/Mimick/blob/e03ab50f348c5df7e4d9866dd988112f0e46ee05/.travis.yml#L38-L40, you see

```yaml
    - env: [CONFIGURATION=Debug, *i386-env]
      addons: *i386-packages
      compiler: gcc
```

This leads to

```json
      {
        "addons": {
          "apt": {
            "packages": [
              "qemu-user", 
              "gcc-arm-linux-gnueabihf"
            ]
          }
        }, 
        "env": [
          "CONFIGURATION=Debug", 
          [
            "ARCH=arm", 
            "TARGET=linux-gnueabihf", 
            "TOOLCHAIN=gcc-generic", 
            "CMAKE_OPTS=-DUSE_QEMU=ON"
          ]
        ], 
        "compiler": "gcc"
      }
```

The `env` is an sequence consisting of a scalar and a sequence. This is invalid, and you get what you see. I believe that you can't merge sequences in YAML, though you can merge maps (http://yaml.org/type/merge.html).

**Comment by Snaipe** [comment]
Right, but travis seems to be flattening the list before running the job, as the builds are loaded with correct environment variables (I can assert this because jobs 1-3 builds a x86_64 asm file while jobs 4-6 builds a i386 asm file, and architecture specification is done through the environment).

I though that the flattening behaviour was expected, as travis does the same for package lists (mentioned in #3505), so I changed the issue title to a display bug.

**Comment by BanzaiMan** [comment]
In its entirety, the configuration looks like this:

```json
{
  "before_install": [
    "if [ -z \"${TARGET}\" ]; then export TARGET=${ARCH}; fi", 
    "if [ \"$TRAVIS_OS_NAME\" = \"osx\" ] && [ \"$CC\" = \"clang\" ]; then\n  brew update\n  brew unlink cmake\n  brew install llvm cmake\nfi\n"
  ], 
  "matrix": {
    "include": [
      {
        "env": "CONFIGURATION=Debug", 
        "compiler": "gcc"
      }, 
      {
        "env": "CONFIGURATION=Release", 
        "compiler": "gcc"
      }, 
      {
        "env": "CONFIGURATION=RelWithDebInfo", 
        "compiler": "gcc"
      }, 
      {
        "addons": {
          "apt": {
            "packages": [
              "gcc-multilib"
            ]
          }
        }, 
        "env": [
          "CONFIGURATION=Debug", 
          [
            "TOOLCHAIN=gcc-i686"
          ]
        ], 
        "compiler": "gcc"
      }, 
      {
        "addons": {
          "apt": {
            "packages": [
              "gcc-multilib"
            ]
          }
        }, 
        "env": [
          "CONFIGURATION=Release", 
          [
            "TOOLCHAIN=gcc-i686"
          ]
        ], 
        "compiler": "gcc"
      }, 
      {
        "addons": {
          "apt": {
            "packages": [
              "gcc-multilib"
            ]
          }
        }, 
        "env": [
          "CONFIGURATION=RelWithDebInfo", 
          [
            "TOOLCHAIN=gcc-i686"
          ]
        ], 
        "compiler": "gcc"
      }, 
      {
        "addons": {
          "apt": {
            "packages": [
              "qemu-user", 
              "gcc-arm-linux-gnueabihf"
            ]
          }
        }, 
        "env": [
          "CONFIGURATION=Debug", 
          [
            "ARCH=arm", 
            "TARGET=linux-gnueabihf", 
            "TOOLCHAIN=gcc-generic", 
            "CMAKE_OPTS=-DUSE_QEMU=ON"
          ]
        ], 
        "compiler": "gcc"
      }, 
      {
        "addons": {
          "apt": {
            "packages": [
              "qemu-user", 
              "gcc-arm-linux-gnueabihf"
            ]
          }
        }, 
        "env": [
          "CONFIGURATION=Release", 
          [
            "ARCH=arm", 
            "TARGET=linux-gnueabihf", 
            "TOOLCHAIN=gcc-generic", 
            "CMAKE_OPTS=-DUSE_QEMU=ON"
          ]
        ], 
        "compiler": "gcc"
      }, 
      {
        "addons": {
          "apt": {
            "packages": [
              "qemu-user", 
              "gcc-arm-linux-gnueabihf"
            ]
          }
        }, 
        "env": [
          "CONFIGURATION=RelWithDebInfo", 
          [
            "ARCH=arm", 
            "TARGET=linux-gnueabihf", 
            "TOOLCHAIN=gcc-generic", 
            "CMAKE_OPTS=-DUSE_QEMU=ON"
          ]
        ], 
        "compiler": "gcc"
      }, 
      {
        "os": "osx", 
        "env": "CONFIGURATION=Debug", 
        "compiler": "clang"
      }, 
      {
        "os": "osx", 
        "env": "CONFIGURATION=Release", 
        "compiler": "clang"
      }, 
      {
        "os": "osx", 
        "env": "CONFIGURATION=RelWithDebInfo", 
        "compiler": "clang"
      }, 
      {
        "os": "osx", 
        "env": [
          "CONFIGURATION=Debug", 
          [
            "TOOLCHAIN=gcc-darwin-i686"
          ]
        ], 
        "compiler": "clang"
      }, 
      {
        "os": "osx", 
        "env": [
          "CONFIGURATION=Release", 
          [
            "TOOLCHAIN=gcc-darwin-i686"
          ]
        ], 
        "compiler": "clang"
      }, 
      {
        "os": "osx", 
        "env": [
          "CONFIGURATION=RelWithDebInfo", 
          [
            "TOOLCHAIN=gcc-darwin-i686"
          ]
        ], 
        "compiler": "clang"
      }
    ]
  }, 
  "language": "c", 
  "script": [
    "mkdir -p build && cd $_", 
    "if [ -n \"${TOOLCHAIN}\" ]; then\n  TC_FILE=\"../.cmake/Toolchains/Toolchain-${TOOLCHAIN}.cmake\"\n  export CMAKE_OPTS=\"${CMAKE_OPTS} -DARCH=${ARCH:-Generic}\"\n  export CMAKE_OPTS=\"${CMAKE_OPTS} -DTARGET=${TARGET:-Any}\"\n  export CMAKE_OPTS=\"${CMAKE_OPTS} -DSYSTEM=Linux\"\n  export CMAKE_OPTS=\"${CMAKE_OPTS} -DCMAKE_TOOLCHAIN_FILE=${TC_FILE}\"\nfi\n", 
    "cmake -Wno-dev -DCMAKE_BUILD_TYPE=${CONFIGURATION} -DCMAKE_INSTALL_PREFIX=mimick-${TRAVIS_TAG} ${CMAKE_OPTS} ..\n", 
    "TERM=dumb cmake --build . -- -j4", 
    "if [ \"${TESTS:-ON}\" = \"ON\" ]; then\n  ctest\nfi"
  ], 
  "_packages": [
    {
      "apt": {
        "packages": [
          "gcc-multilib"
        ]
      }
    }, 
    {
      "apt": {
        "packages": [
          "qemu-user", 
          "gcc-arm-linux-gnueabihf"
        ]
      }
    }
  ], 
  "sudo": false, 
  "_environments": [
    [
      "TOOLCHAIN=gcc-i686"
    ], 
    [
      "TOOLCHAIN=gcc-darwin-i686"
    ], 
    [
      "ARCH=arm", 
      "TARGET=linux-gnueabihf", 
      "TOOLCHAIN=gcc-generic", 
      "CMAKE_OPTS=-DUSE_QEMU=ON"
    ]
  ]
}
```

You will see that, when `env` is a scalar or a simple sequence, the jobs are configured correctly. In other cases your jobs fail.

**Comment by Snaipe** [comment]
I'm using almost the [same config](https://github.com/Snaipe/Mimick/blob/10ad761c9e730f2b4120118544e071a19c5b1ce1/.travis.yml) right now, and [all jobs pass](https://travis-ci.org/Snaipe/Mimick/builds/120427475) -- the build failures were unrelated at the time I posted this, and all jobs have valid environment variables (1-3 logs show valid x64 compilation, 4-6 i386, and 7-9 arm)

The env key always gets flattened as one sequence, and while I don't know if it is more of a consequence of the implementation (as it seems to be undocumented), there is a disconnect between the code that displays the summary line (which doesn't handle complex sequences), and the one that loads the environment (which does).

Perhaps this issue should be "Officially allow and document nested sequences in the .travis.yml env key".

**Comment by Snaipe** [comment]
> The env key always gets flattened as one sequence

actually, let me rephrase that: the env key might not seem to be flattened, but visibly *appears* to be since nested sequences are "correctly" handled by the environment loader.

**Comment by BanzaiMan** [comment]
It was incorrect to say that the jobs fail. I apologize for that.

I was under the impression that this issue was about how the environment variables are displayed in the build page. Am I wrong?

We should normalize config data to some extent, and I believe we do the environment variables in the way you are asking. The display on the web page, though, will not go through the normalization, so you get what you see.

**Comment by Snaipe** [comment]
> I was under the impression that this issue was about how the environment variables are displayed in the build page. Am I wrong?

No, you're right -- what I was saying earlier is that the normalization process for the `env` key isn't documented, so I wasn't sure if it was really supported.

In all cases, the display issue remains.

**Comment by PeterDaveHello** [comment]
Do we have updates on this problem? Thanks :)
